File: backend/app/lib/parser.py
"""
Service notation parser — ported from the original TypeScript parser.ts.

Parses strings like "80 + 油5 + 淋20 + 拔20 + 美35 + 刮10 + T15" into
structured data.
"""

import re
from typing import Any

# ── Service code definitions (keep in sync with seed data) ──

SERVICE_CODES: list[dict[str, Any]] = [
    {"name": "Massage", "chinese_name": "按摩", "aliases": [], "default_amount": None, "common_amounts": [], "category": "massage", "min_amount": None, "max_amount": None},
    {"name": "Oil Upgrade", "chinese_name": "油", "aliases": ["油", "O", "Oil"], "default_amount": 5, "common_amounts": [5, 10], "category": "massage", "min_amount": 0, "max_amount": 30},
    {"name": "CBD Oil", "chinese_name": "CBD油", "aliases": ["CBD"], "default_amount": 20, "common_amounts": [20], "category": "massage", "min_amount": 0, "max_amount": 40},
    {"name": "Lymphatic", "chinese_name": "淋巴", "aliases": ["淋", "L", "LY"], "default_amount": 20, "common_amounts": [20], "category": "massage", "min_amount": 0, "max_amount": 40},
    {"name": "Cupping", "chinese_name": "拔罐", "aliases": ["拔", "C", "CUP"], "default_amount": 20, "common_amounts": [20], "category": "massage", "min_amount": 0, "max_amount": 40},
    {"name": "Facial", "chinese_name": "美容", "aliases": ["美", "F", "FACIAL"], "default_amount": 35, "common_amounts": [35], "category": "facial", "min_amount": 0, "max_amount": 100},
    {"name": "Gua Sha", "chinese_name": "刮痧", "aliases": ["刮", "G", "GS"], "default_amount": 10, "common_amounts": [10, 40], "category": "massage", "min_amount": 0, "max_amount": 60},
    {"name": "Muscle Gun", "chinese_name": "筋膜枪", "aliases": ["枪", "MG"], "default_amount": 15, "common_amounts": [15], "category": "massage", "min_amount": 0, "max_amount": 30},
    {"name": "Thai Stretch", "chinese_name": "泰式拉伸", "aliases": ["拉", "TS"], "default_amount": 15, "common_amounts": [15], "category": "massage", "min_amount": 0, "max_amount": 30},
    {"name": "Card Tip", "chinese_name": "刷卡小费", "aliases": ["T", "Tip"], "default_amount": None, "common_amounts": [], "category": "card_tip", "min_amount": 0, "max_amount": 500},
    {"name": "Cash Tip", "chinese_name": "现金小费", "aliases": ["CT", "CashTip"], "default_amount": None, "common_amounts": [], "category": "cash_tip", "min_amount": 0, "max_amount": 500},
]

# Build alias map: alias -> code definition
_ALIAS_MAP: dict[str, dict[str, Any]] = {}
for code in SERVICE_CODES:
    _ALIAS_MAP[code["name"].lower()] = code
    _ALIAS_MAP[code["chinese_name"]] = code
    for alias in code["aliases"]:
        _ALIAS_MAP[alias.lower()] = code


def normalize_separators(text: str) -> str:
    """Normalize Chinese commas, slashes, whitespace to +."""
    text = text.replace("，", ",").replace("、", ",")
    text = re.sub(r"[/\\]", "+", text)
    text = re.sub(r"\s*,\s*", "+", text)
    text = re.sub(r"\s*\+\s*", "+", text)
    text = re.sub(r"\s+", "+", text)
    text = re.sub(r"^\+", "", text)
    text = re.sub(r"\+$", "", text)
    text = re.sub(r"\++", "+", text)
    return text


def match_alias(token: str) -> dict[str, Any] | None:
    """Try to match a token against the alias map. Returns the code def or None."""
    lower = token.lower()
    # Exact match first
    if lower in _ALIAS_MAP:
        return _ALIAS_MAP[lower]

    # Sort by longest alias first
    entries = sorted(_ALIAS_MAP.items(), key=lambda x: -len(x[0]))
    for alias, code in entries:
        if lower.startswith(alias):
            return code

    return None


def is_numeric(s: str) -> bool:
    return bool(re.match(r"^-?\d+(\.\d+)?$", s))


def parse_service_notation(raw_text: str) -> dict[str, Any]:
    """
    Parse a service notation string into structured data.

    Args:
        raw_text: e.g. "80 + 油5 + 淋20 + 拔20 + 美35 + 刮10 + T15"

    Returns:
        dict with parsed fields
    """
    warnings: list[str] = []
    unknown_tokens: list[str] = []
    items: list[dict[str, Any]] = []
    massage_base = 0.0
    card_tip = 0.0
    cash_tip = 0.0

    normalized = normalize_separators(raw_text.strip())

    if not normalized:
        return {
            "raw_text": raw_text,
            "massage_base": 0,
            "items": [],
            "card_tip": 0,
            "cash_tip": 0,
            "massage_total": 0,
            "facial_total": 0,
            "service_total": 0,
            "total_with_tips": 0,
            "unknown_tokens": [],
            "warnings": ["原始记账内容为空"],
        }

    tokens = [t.strip() for t in normalized.split("+") if t.strip()]
    found_massage_base = False

    for token in tokens:
        # Try alias + amount (e.g. "油5", "美35.5", "T15")
        m = re.match(r"^([a-zA-Z\u4e00-\u9fff]+)(\d+(?:\.\d+)?)$", token)
        if m:
            alias, amount_str = m.groups()
            amount = float(amount_str)
            code_def = match_alias(alias)

            if code_def and code_def["category"] == "card_tip":
                card_tip += amount
                items.append({"code": code_def["chinese_name"], "name": code_def["name"], "amount": amount, "category": "card_tip", "original_token": token, "confidence": 1.0})
                continue
            if code_def and code_def["category"] == "cash_tip":
                cash_tip += amount
                items.append({"code": code_def["chinese_name"], "name": code_def["name"], "amount": amount, "category": "cash_tip", "original_token": token, "confidence": 1.0})
                continue
            if code_def:
                items.append({"code": code_def["chinese_name"], "name": code_def["name"], "amount": amount, "category": code_def["category"], "original_token": token, "confidence": 1.0})
                if code_def["max_amount"] is not None and amount > code_def["max_amount"]:
                    warnings.append(f"{code_def['name']}金额为 ${amount}，可能异常（上限 ${code_def['max_amount']}）")
                if code_def["min_amount"] is not None and amount < code_def["min_amount"]:
                    warnings.append(f"{code_def['name']}金额为 ${amount}，可能异常（下限 ${code_def['min_amount']}）")
                continue
            else:
                unknown_tokens.append(token)
                warnings.append(f"不认识的项目缩写: {alias}")
                continue

        # Try bare number (massage base)
        if is_numeric(token) and not found_massage_base:
            massage_base = float(token)
            found_massage_base = True
            continue

        # Try alias only (no amount)
        if re.match(r"^[a-zA-Z\u4e00-\u9fff]+$", token):
            code_def = match_alias(token)
            if code_def:
                default_amount = code_def["default_amount"] or 0
                if code_def["category"] == "card_tip":
                    card_tip += default_amount
                elif code_def["category"] == "cash_tip":
                    cash_tip += default_amount
                else:
                    items.append({"code": code_def["chinese_name"], "name": code_def["name"], "amount": default_amount, "category": code_def["category"], "original_token": token, "confidence": 1.0})
                continue

        # Second number is suspicious
        if is_numeric(token) and found_massage_base:
            unknown_tokens.append(token)
            warnings.append(f"发现第二个数字 {token}，可能缺少项目缩写")
            continue

        unknown_tokens.append(token)
        warnings.append(f"无法识别: {token}")

    if not found_massage_base and massage_base == 0:
        warnings.append("未找到按摩基础金额")

    massage_total = massage_base + sum(i["amount"] for i in items if i["category"] == "massage")
    facial_total = sum(i["amount"] for i in items if i["category"] == "facial")
    service_total = massage_total + facial_total
    total_with_tips = service_total + card_tip + cash_tip

    # Check duplicates
    seen = set()
    for item in items:
        if item["code"] in seen:
            warnings.append(f"项目 {item['code']} 可能重复")
        seen.add(item["code"])

    if card_tip > service_total:
        warnings.append("小费高于服务金额")

    return {
        "raw_text": raw_text,
        "massage_base": massage_base,
        "items": items,
        "card_tip": card_tip,
        "cash_tip": cash_tip,
        "massage_total": massage_total,
        "facial_total": facial_total,
        "service_total": service_total,
        "total_with_tips": total_with_tips,
        "unknown_tokens": unknown_tokens,
        "warnings": warnings,
    }

File: work-log/src/lib/parser.ts
import { ParsedNotation, ParsedItem, ServiceCodeDef } from "@/types";
import { DEFAULT_SERVICE_CODES, buildAliasMap } from "./service-codes";

const aliasMap = buildAliasMap(DEFAULT_SERVICE_CODES);

/**
 * Normalize separators: +, ，, ,, /, \, multiple spaces, newlines → "+"
 */
function normalizeSeparators(text: string): string {
  return text
    .replace(/[，、]/g, ",")       // Chinese commas → English comma
    .replace(/[\/\\]/g, "+")       // slashes → +
    .replace(/\s*,\s*/g, "+")      // comma with spaces → +
    .replace(/\s*\+\s*/g, "+")     // normalize spaces around +
    .replace(/\s+/g, "+")          // remaining whitespace → +
    .replace(/^\+/, "")            // leading +
    .replace(/\+$/, "")            // trailing +
    .replace(/\++/g, "+");         // collapse multiple +
}

/**
 * Try to match a token against the alias map.
 * Returns the matched ServiceCodeDef or null.
 * Prefers longer aliases to avoid partial matches (e.g. "CBD" vs "C").
 */
function matchAlias(token: string): { code: ServiceCodeDef; matchedAlias: string } | null {
  const lower = token.toLowerCase();
  // Check exact match first
  const exact = aliasMap.get(lower);
  if (exact) return { code: exact, matchedAlias: token };

  // Build sorted entries: longest alias first to avoid partial matches
  const entries = Array.from(aliasMap.entries()).sort(
    (a, b) => b[0].length - a[0].length
  );

  for (const [alias, code] of entries) {
    if (lower.startsWith(alias)) {
      return { code, matchedAlias: token.slice(0, alias.length) };
    }
  }

  return null;
}

/**
 * Check if a string is purely numeric (integer or decimal)
 */
function isNumeric(s: string): boolean {
  return /^-?\d+(\.\d+)?$/.test(s);
}

/**
 * Parse a service notation string into structured data.
 *
 * @param rawText - The raw text from OCR or manual input
 * @returns ParsedNotation with all fields
 *
 * @example
 * parseServiceNotation("80 + 油5 + 淋20 + 拔20 + 美35 + 刮10 + T15")
 */
export function parseServiceNotation(rawText: string): ParsedNotation {
  const warnings: string[] = [];
  const unknownTokens: string[] = [];
  const items: ParsedItem[] = [];
  let massageBase = 0;
  let cardTip = 0;
  let cashTip = 0;

  // Step 1: Normalize
  const normalized = normalizeSeparators(rawText.trim());

  if (!normalized) {
    return {
      rawText,
      massageBase: 0,
      items: [],
      cardTip: 0,
      cashTip: 0,
      massageTotal: 0,
      facialTotal: 0,
      serviceTotal: 0,
      totalWithTips: 0,
      unknownTokens: [],
      warnings: ["原始记账内容为空"],
    };
  }

  // Step 2: Split by +
  const tokens = normalized.split("+").filter((t) => t.length > 0);

  let foundMassageBase = false;

  for (const token of tokens) {
    const trimmed = token.trim();

    // Try to match as alias + amount (e.g. "油5", "美35.5", "T15")
    const match = trimmed.match(/^([a-zA-Z\u4e00-\u9fff]+)(\d+(?:\.\d+)?)$/);
    if (match) {
      const [, alias, amountStr] = match;
      const amount = parseFloat(amountStr);

      // Check if it's a tip
      const tipMatch = matchAlias(alias);
      if (tipMatch && tipMatch.code.category === "card_tip") {
        cardTip += amount;
        items.push({
          code: tipMatch.code.chineseName,
          name: tipMatch.code.name,
          amount,
          category: "card_tip",
          originalToken: trimmed,
          confidence: 1.0,
        });
        continue;
      }
      if (tipMatch && tipMatch.code.category === "cash_tip") {
        cashTip += amount;
        items.push({
          code: tipMatch.code.chineseName,
          name: tipMatch.code.name,
          amount,
          category: "cash_tip",
          originalToken: trimmed,
          confidence: 1.0,
        });
        continue;
      }

      // Regular service item
      if (tipMatch) {
        items.push({
          code: tipMatch.code.chineseName,
          name: tipMatch.code.name,
          amount,
          category: tipMatch.code.category as "massage" | "facial",
          originalToken: trimmed,
          confidence: 1.0,
        });

        // Check amount warnings
        if (tipMatch.code.maxAmount !== undefined && amount > tipMatch.code.maxAmount) {
          warnings.push(`${tipMatch.code.name}金额为 $${amount}，可能异常（上限 $${tipMatch.code.maxAmount}）`);
        }
        if (tipMatch.code.minAmount !== undefined && amount < tipMatch.code.minAmount) {
          warnings.push(`${tipMatch.code.name}金额为 $${amount}，可能异常（下限 $${tipMatch.code.minAmount}）`);
        }
      } else {
        unknownTokens.push(trimmed);
        warnings.push(`不认识的项目缩写: ${alias}`);
      }
      continue;
    }

    // Try to match as a number first (massage base)
    if (isNumeric(trimmed) && !foundMassageBase) {
      massageBase = parseFloat(trimmed);
      foundMassageBase = true;
      continue;
    }

    // Try to match as alias only (no amount, use default)
    if (/^[a-zA-Z\u4e00-\u9fff]+$/.test(trimmed)) {
      const tipMatch = matchAlias(trimmed);
      if (tipMatch) {
        const defaultAmount = tipMatch.code.defaultAmount || 0;
        if (tipMatch.code.category === "card_tip") {
          cardTip += defaultAmount;
        } else if (tipMatch.code.category === "cash_tip") {
          cashTip += defaultAmount;
        } else {
          items.push({
            code: tipMatch.code.chineseName,
            name: tipMatch.code.name,
            amount: defaultAmount,
            category: tipMatch.code.category as "massage" | "facial",
            originalToken: trimmed,
            confidence: 1.0,
          });
        }
        continue;
      }
    }

    // If we already found a base amount and this is a number, it's suspicious
    if (isNumeric(trimmed) && foundMassageBase) {
      unknownTokens.push(trimmed);
      warnings.push(`发现第二个数字 ${trimmed}，可能缺少项目缩写`);
      continue;
    }

    // Unknown token
    unknownTokens.push(trimmed);
    warnings.push(`无法识别: ${trimmed}`);
  }

  // Check for missing massage base
  if (!foundMassageBase && massageBase === 0) {
    warnings.push("未找到按摩基础金额");
  }

  // Calculate totals
  const massageTotal = massageBase + items
    .filter((i) => i.category === "massage")
    .reduce((sum, i) => sum + i.amount, 0);

  const facialTotal = items
    .filter((i) => i.category === "facial")
    .reduce((sum, i) => sum + i.amount, 0);

  const serviceTotal = massageTotal + facialTotal;
  const totalWithTips = serviceTotal + cardTip + cashTip;

  // Check for duplicate items
  const itemCodes = items.map((i) => i.code);
  const seen = new Set<string>();
  for (const code of itemCodes) {
    if (seen.has(code)) {
      warnings.push(`项目 ${code} 可能重复`);
    }
    seen.add(code);
  }

  // Check tip warnings
  if (cardTip > serviceTotal) {
    warnings.push("小费高于服务金额");
  }

  return {
    rawText,
    massageBase,
    items,
    cardTip,
    cashTip,
    massageTotal,
    facialTotal,
    serviceTotal,
    totalWithTips,
    unknownTokens,
    warnings,
  };
}

/**
 * Format structured data back into standard notation string.
 *
 * @example
 * formatServiceNotation({ massageBase: 80, items: [...], cardTip: 15, cashTip: 0 })
 * // => "80 + 油5 + 淋20 + 拔20 + 美35 + 刮10 + T15"
 */
export function formatServiceNotation(data: {
  massageBase: number;
  items: { code: string; amount: number; category: string }[];
  cardTip: number;
  cashTip: number;
}): string {
  const parts: string[] = [];

  if (data.massageBase > 0) {
    parts.push(String(data.massageBase));
  }

  for (const item of data.items) {
    if (item.category === "card_tip" || item.category === "cash_tip") continue;
    // Find the code definition and use the shortest alias (the abbreviation)
    const codeDef = DEFAULT_SERVICE_CODES.find(
      (c) => c.name === item.code || c.chineseName === item.code
    );
    // Use the first alias (short code) or chineseName as the display code
    const code = codeDef ? (codeDef.aliases[0] || codeDef.chineseName) : item.code;
    parts.push(`${code}${item.amount}`);
  }

  if (data.cardTip > 0) {
    parts.push(`T${data.cardTip}`);
  }

  if (data.cashTip > 0) {
    parts.push(`CT${data.cashTip}`);
  }

  return parts.join(" + ");
}